// vue/dist/vue.esm-bundler.js
/**
  * vue v3.6.0-beta.5
  * (c) 2018-present Yuxi (Evan) You and Vue contributors
  * @license MIT
  **/
import * as runtimeDom from "@vue/runtime-dom";
import { initCustomFormatter, registerRuntimeCompiler, warn } from "@vue/runtime-dom";
import { compile } from "@vue/compiler-dom";
import { NOOP, extend, genCacheKey, generateCodeFrame, isString } from "@vue/shared";

export * from "@vue/runtime-dom"

export * from "@vue/runtime-vapor"

//#region \0rolldown/runtime.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __exportAll = (all, no_symbols) => {
	let target = {};
	for (var name in all) {
		__defProp(target, name, {
			get: all[name],
			enumerable: true
		});
	}
	if (!no_symbols) {
		__defProp(target, Symbol.toStringTag, { value: "Module" });
	}
	return target;
};
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") {
		for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
			key = keys[i];
			if (!__hasOwnProp.call(to, key) && key !== except) {
				__defProp(to, key, {
					get: ((k) => from[k]).bind(null, key),
					enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
				});
			}
		}
	}
	return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));

//#endregion
//#region packages/vue/src/dev.ts
function initDev() {
	initCustomFormatter();
}

//#endregion
//#region packages/vue/src/index.ts
if (!!(process.env.NODE_ENV !== "production")) initDev();
const compileCache = Object.create(null);
function compileToFunction(template, options) {
	if (!isString(template)) if (template.nodeType) template = template.innerHTML;
	else {
		process.env.NODE_ENV !== "production" && warn(`invalid template option: `, template);
		return NOOP;
	}
	const key = genCacheKey(template, options);
	const cached = compileCache[key];
	if (cached) return cached;
	if (template[0] === "#") {
		const el = document.querySelector(template);
		if (!!(process.env.NODE_ENV !== "production") && !el) warn(`Template element not found or is empty: ${template}`);
		template = el ? el.innerHTML : ``;
	}
	const opts = extend({
		hoistStatic: true,
		onError: !!(process.env.NODE_ENV !== "production") ? onError : void 0,
		onWarn: !!(process.env.NODE_ENV !== "production") ? (e) => onError(e, true) : NOOP
	}, options);
	if (!opts.isCustomElement && typeof customElements !== "undefined") opts.isCustomElement = (tag) => !!customElements.get(tag);
	const { code } = compile(template, opts);
	function onError(err, asWarning = false) {
		const message = asWarning ? err.message : `Template compilation error: ${err.message}`;
		const codeFrame = err.loc && generateCodeFrame(template, err.loc.start.offset, err.loc.end.offset);
		warn(codeFrame ? `${message}\n${codeFrame}` : message);
	}
	const render = new Function("Vue", code)(runtimeDom);
	render._rc = true;
	return compileCache[key] = render;
}
registerRuntimeCompiler(compileToFunction);

//#endregion
export { compileToFunction as compile };
